pkg/azureutils/azure_disk_utils.go

```diff
@@ -1090,68 +1090,29 @@ func GetAzVolumeAttachmentState(volumeAttachmentStatus storagev1.VolumeAttachmen
 
 type UpdateCRIFunc func(client.Object) error
 
-func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.SharedInformerFactory, cachedClient client.Client, azDiskClient azdisk.Interface, obj client.Object, updateFunc UpdateCRIFunc, maxNetRetry int, updateMode CRIUpdateMode) (client.Object, error) {
+func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.SharedInformerFactory, cachedClient client.Client, azDiskClient azdisk.Interface, obj client.Object, originalObj client.Object, updateFunc UpdateCRIFunc, maxNetRetry int, updateMode CRIUpdateMode) (client.Object, error) {
```

Comment:
Why add `originalObj` when we already pass the original object in `obj`?

Reply:
I agree that `obj` and `originalObj` contain duplicated data.

However, `obj` can be the original object or updated object. This can provide some flexibility to call `UpdateCRIWithRetry()`. So, when we loss the original object we can pass nil.

For example, inside `ReconcileAttachDetach.updateVolumeAttachmentWithResult()`, we don't have original object.
```
va := storagev1.VolumeAttachment{ObjectMeta: metav1.ObjectMeta{Name: vaName}}
_, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, &va, nil, vaUpdateFunc, consts.ForcedUpdateMaxNetRetry, azureutils.UpdateCRIStatus)

 ```

In addition, in `CrdProvisioner.PublishVolume()`, it seems also loss the original `attachmentObj`, because this line: https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/main_v2/pkg/provisioner/crdprovisioner.go#L591
is updating `attachmentObj.Annotations`

Also, inside `CrdProvisioner.UnpulishVolume()`, it will first `demoteVolume(ctx, azVolumeAttachmentInstance)`, then `detachVolume(ctx, azVolumeAttachmentInstance)`. `azVolumeAttachmentInstance` might be updated, and we loss the original object. For this case, it passes nil to `originalObj`, which can force `client.Get()` being called to get the original object.

I was thinking that we can pass `objInfo{Namespace, Name, Type}` instead of obj to reduce the duplicated data passing. But not sure if it worths or not.

Reply:
BTW, it's fine to have `originalObj` defined as a local variable and used to track the original object. This is needed to prevent unnecessary PUT calls to API server if the state of the object doesn't change. However, it should start out as a copy of the input `obj` and refreshed by GET on conflict.